File: 2018/03/fabric.cc
#include <cassert>
#include <fstream>
#include <iostream>
#include <set>
#include <sstream>
#include <string>
#include <vector>

#include <re2/re2.h>

using namespace std; 

class Point {
public:
	int	x;
	int	y;
	Point() : x(0), y(0) {};
	Point(int _x, int _y) : x(_x), y(_y) {};
};

inline bool
operator==(const Point &lhs, const Point &rhs)
{
	return lhs.x == rhs.x && lhs.y == rhs.y;
}

inline bool
operator!=(const Point &lhs, const Point& rhs)
{
	return !(lhs == rhs);
}

inline bool
operator<(const Point &lhs, const Point& rhs)
{
	if (lhs.y < rhs.y) {
		return true;
	}

	if (lhs.y > rhs.y) {
		return false;
	}

	if (lhs.x < rhs.x) {
		return true;
	}

	return false;
}

ostream&
operator<<(ostream& os, const Point& point)
{
	os << "(" << point.x << ", " << point.y << ")";
	return os;
}

static string	claim_regex_str = "^#(\\d+) @ (\\d+),(\\d+): (\\d+)x(\\d+)$";
static re2::RE2	claimre(claim_regex_str);

class Claim{
public:
	int	id;
	Point	upperLeft;
	Point	lowerRight;

	Claim() : id(0), upperLeft(Point(0, 0)), lowerRight(Point(0, 0)) {};
	Claim(int id, Point upperLeft, Point upperRight) : id(id), upperLeft(upperLeft), lowerRight(upperRight) {};
	void dump() { cout << "#" << id << ": (" << upperLeft.x << ", " << upperLeft.y << "), (" << lowerRight.x << ", " << lowerRight.y << ")\n"; }
};

inline bool
operator==(const Claim &lhs, const Claim &rhs)
{
	return (lhs.id == rhs.id) && (lhs.upperLeft == rhs.upperLeft) && (lhs.lowerRight == rhs.lowerRight);
}

inline const bool
operator!=(const Claim &lhs, const Claim& rhs)
{
	return !(lhs == rhs);
}

ostream&
operator<<(ostream &os, const Claim &claim)
{
	os << "#" << claim.id << " " << claim.upperLeft << ", " << claim.lowerRight;
	return os;
}

static Claim
process_claim(string claimString)
{
	Claim	claim;

	RE2::FullMatch(claimString, claimre, &claim.id, &claim.upperLeft.x,
	    &claim.upperLeft.y, &claim.lowerRight.x, &claim.lowerRight.y);
	claim.lowerRight.x += claim.upperLeft.x;
	claim.lowerRight.y += claim.upperLeft.y;

	return claim;
}

static void
self_check_process_claim()
{
	auto claimString = "#1 @ 257,829: 10x23";
	Claim expected(1, Point(257, 829), Point(267, 852));
	auto claim = process_claim(claimString);
	assert(claim == expected);
}

static void
self_check_overlapping_claim()
{
	self_check_process_claim();
}

static int
count_claim(Claim &claim, set<Point> &counter)
{
	int	overlapping = 0;

	for (auto j = claim.upperLeft.y; j < claim.lowerRight.y; j++) {
		for (auto i = claim.upperLeft.x; i < claim.lowerRight.x; i++) {
			auto point = Point(i, j);
			counter.insert(point);
			if (counter.count(point) == 2) {
				overlapping++;
			}
		}
	}

	return overlapping;
}

static int
count_claims(vector<Claim> &claims, set<Point> &counter)
{
	int	overlapping = 0;

	for (auto claim : claims) {
		overlapping += count_claim(claim, counter);
	}
	cout << "counter has seen " << counter.size() << " points." << endl;
	return overlapping;
}	

static void
self_check()
{
	// Part 1
	self_check_overlapping_claim();
}

static vector<Claim>
load_claims(string path)
{
	ifstream	claim_file(path);
	vector<Claim>	claims;

	for (string line; getline(claim_file, line); ) {
		claims.push_back(process_claim(line));
	}
	claim_file.close();

	return claims;
}

int
main(int argc, char *argv[])
{
	self_check();
	for (auto i = 1; i < argc; i++) {
		auto claims = load_claims(argv[i]);
		assert(claims.size() > 0);
		cout << "loaded " << claims.size() << " claims." << endl;

		set<Point> counter;
		auto count = count_claims(claims, counter);
		cout << "overlapping area: " << count << endl;
	}
}

File: 2018/03/fabric.py
#!/usr/bin/env python3

import collections
import re
import sys
from typing import List


class Claim:
    __slots__ = ["id", "upper_left", "lower_right"]

    def __init__(self, id: int, x0: int, y0: int, x1: int, y1: int):
        self.id = int(id)
        self.upper_left = (x0, y0)
        self.lower_right = (x1, y1)

    def __eq__(self, other):
        if self.id != other.id:
            return False
        if self.upper_left != other.upper_left:
            return False
        if self.lower_right != other.lower_right:
            return False
        return True

    def points(self):
        points = []
        for j in range(self.upper_left[1], self.lower_right[1]):
            for i in range(self.upper_left[0], self.lower_right[0]):
                points.append((i, j))
        return points


def create_claim_regex():
    return re.compile(r"^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$")


claim_regex = create_claim_regex()


def process_claim(claim: str) -> Claim:
    m = claim_regex.match(claim.strip())
    x0 = int(m.group(2))
    y0 = int(m.group(3))
    x1 = x0 + int(m.group(4))
    y1 = y0 + int(m.group(5))
    return Claim(m.group(1), x0, y0, x1, y1)


def self_check_process_claim():
    claim_strings = ["#7 @ 922,250: 13x26", "#8 @ 256,742: 18x14"]
    expected = [Claim(7, 922, 250, 935, 276), Claim(8, 256, 742, 274, 756)]

    for i in range(len(claim_strings)):
        claim = process_claim(claim_strings[i])
        assert claim == expected[i]


def load_claims(path: str) -> List[Claim]:
    with open(path, "rt") as claim_file:
        return list(map(process_claim, claim_file.readlines()))


def map_points(claims):
    counter = collections.Counter()
    for claim in claims:
        points = claim.points()
        for point in points:
            counter[point] += 1

    return counter


def overlapping(points):
    return len([pt for pt in points.keys() if points[pt] > 1])


def alone(claims, counter):
    for claim in claims:
        non_overlapping = 0
        points = claim.points()
        for point in points:
            if counter[point] == 1:
                non_overlapping += 1
        if len(points) == non_overlapping:
            return claim.id
    return None


def self_check_overlapping():
    claims = list(
        map(process_claim, ["#1 @ 1,3: 4x4", "#2 @ 3,1: 4x4", "#3 @ 5,5: 2x2"])
    )
    area = overlapping(map_points(claims))
    assert area == 4


def self_check():
    self_check_process_claim()
    # self_check_points()
    self_check_overlapping
    print("self check OK")


def main(args):
    for arg in args:
        claims = load_claims(arg)
        print("Loaded {} claims".format(len(claims)))
        counter = map_points(claims)
        area = overlapping(counter)
        print("Overlapping area:", area)

        lone = alone(claims, counter)
        print("Lone claim:", lone)


if __name__ == "__main__":
    self_check()
    main(sys.argv[1:])
